internal/training/grpo_engine/kernels.c

```
@@ -0,0 +1,764 @@
+        const float *qi = q + h * head_dim;
+
+        /* Compute attention scores for this head */
+        float *scores = (float *)malloc((seq_pos + 1) * sizeof(float));
+        float scale = 1.0f / sqrtf((float)head_dim);
+
+        for (int t = 0; t <= seq_pos; t++) {
+            const float *kt = k_cache + (size_t)t * n_kv_heads * head_dim + kv_h * head_dim;
+            double dot = 0.0;
+            for (int d = 0; d < head_dim; d++)
+                dot += (double)qi[d] * (double)kt[d];
+            scores[t] = (float)dot * scale;
```

Cppcheck (2.21.0)

[warning] If memory allocation fails, then there is a possible null pointer dereference

(nullPointerOutOfMemory)